File: Frontend/src/types/goal.ts
export interface Goal {
  id: number;
  title: string;
  description: string;
  status: 'completed' | 'on-track' | 'in-progress' | 'overdue' | 'paused';
  category: string;
  deadline: string;
  priority: 'high' | 'medium' | 'low';
  created_at: string;
  updated_at: string;
}

export interface CreateGoalRequest {
  title: string;
  description: string;
  status?: Goal['status'];
  category: string;
  deadline?: string;
  priority: Goal['priority'];
}

export interface UpdateGoalRequest {
  title?: string;
  description?: string;
  status?: Goal['status'];
  category?: string;
  deadline?: string;
  priority?: Goal['priority'];
}

export interface GoalStats {
  total: number;
  inProgress: number;
  onTrack: number;
  atRisk: number;
}

export interface GoalListResponse {
  data: Goal[];
  total: number;
  page: number;
  page_size: number;
  message: string;
}

export interface GoalResponse {
  data: Goal;
  message: string;
}

export interface GoalStatsResponse {
  data: {
    total_goals: number;
    completed_goals: number;
    in_progress_goals: number;
    overdue_goals: number;
  };
  message: string;
}

export interface GoalFilters {
  page?: number;
  page_size?: number;
  category?: string;
  status?: Goal['status'];
  priority?: Goal['priority'];
  search?: string;
}

File: Backend/app/schemas/goal.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from enum import Enum


class GoalStatus(str, Enum):
    in_progress = "in-progress"
    on_track = "on-track"
    completed = "completed"
    overdue = "overdue"
    paused = "paused"


class GoalPriority(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"


class GoalBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    status: GoalStatus = GoalStatus.in_progress
    category: str = Field(..., min_length=1, max_length=50)
    deadline: Optional[str] = None  # ISO date string format
    priority: GoalPriority = GoalPriority.medium


class GoalCreate(GoalBase):
    pass


class GoalUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    status: Optional[GoalStatus] = None
    category: Optional[str] = Field(None, min_length=1, max_length=50)
    deadline: Optional[str] = None
    priority: Optional[GoalPriority] = None


class Goal(GoalBase):
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class GoalResponse(BaseModel):
    data: Goal
    message: str = "Success"


class GoalsListResponse(BaseModel):
    data: list[Goal]
    total: int
    page: int
    page_size: int
    message: str = "Success"
